src/mapping/query/functions.rs

```diff
@@ -205,79 +472,137 @@ fn to_timestamp(value: Value) -> Result<Value> {
 
 #[derive(Debug)]
 pub(in crate::mapping) struct ParseTimestampFn {
-    conversion: Conversion,
     query: Box<dyn Function>,
-    default: Option<Value>,
+    format: Box<dyn Function>,
+    default: Option<Box<dyn Function>>,
 }
 
 impl ParseTimestampFn {
+    #[cfg(test)]
     pub(in crate::mapping) fn new(
         format: &str,
         query: Box<dyn Function>,
         default: Option<Value>,
-    ) -> Result<Self> {
-        let conversion: Conversion = ("timestamp|".to_string() + format)
-            .parse()
-            .map_err(|e| format!("{}", e))?;
-        Ok(Self {
-            conversion,
+    ) -> Self {
+        Self {
             query,
-            default,
-        })
+            format: Box::new(Literal::from(Value::from(format))),
+            default: default.map(|v| Box::new(Literal::from(v)) as _),
+        }
     }
 }
 
 impl Function for ParseTimestampFn {
     fn execute(&self, ctx: &Event) -> Result<Value> {
+        let format = match self.format.execute(ctx)? {
+            Value::Bytes(b) => format!("timestamp|{}", String::from_utf8_lossy(&b)),
+            v => unexpected_type!(v),
+        };
+
+        let conversion: Conversion = format.parse().map_err(|e| format!("{}", e))?;
```

Comment:
This is an example of computation moving from compile– to runtime. I think this is fine for now, but as mentioned, we could downcast to `Literal` and create the `Conversion` at compile time, if possible.